File: go/indicators/core/lineindicator.go
package core

import (
	"zpano/entities"
)

// LineIndicator provides UpdateScalar, UpdateBar, UpdateQuote and UpdateTrade
// methods for indicators that take a single numeric input and produce a single
// scalar output.
//
// Embed this struct in a concrete indicator and initialize it using
// NewLineIndicator, passing the concrete indicator's Update method as updateFn.
// The four Update* methods will be promoted to the embedding struct and satisfy
// the Indicator interface.
type LineIndicator struct {
	// Mnemonic is a short name of the indicator.
	Mnemonic string

	// Description is a description of the indicator.
	Description string

	barFunc   entities.BarFunc
	quoteFunc entities.QuoteFunc
	tradeFunc entities.TradeFunc
	updateFn  func(float64) float64
}

// NewLineIndicator creates a LineIndicator with the given component functions
// and an update function that computes the indicator value from a single sample.
func NewLineIndicator(
	mnemonic, description string,
	barFunc entities.BarFunc, quoteFunc entities.QuoteFunc, tradeFunc entities.TradeFunc,
	updateFn func(float64) float64,
) LineIndicator {
	return LineIndicator{
		Mnemonic:    mnemonic,
		Description: description,
		barFunc:     barFunc,
		quoteFunc:   quoteFunc,
		tradeFunc:   tradeFunc,
		updateFn:    updateFn,
	}
}

// UpdateScalar updates the indicator given the next scalar sample.
func (l *LineIndicator) UpdateScalar(sample *entities.Scalar) Output {
	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: l.updateFn(sample.Value)}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (l *LineIndicator) UpdateBar(sample *entities.Bar) Output {
	return l.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: l.barFunc(sample)})
}

// UpdateQuote updates the indicator given the next quote sample.
func (l *LineIndicator) UpdateQuote(sample *entities.Quote) Output {
	return l.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: l.quoteFunc(sample)})
}

// UpdateTrade updates the indicator given the next trade sample.
func (l *LineIndicator) UpdateTrade(sample *entities.Trade) Output {
	return l.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: l.tradeFunc(sample)})
}

File: py/indicators/core/line_indicator.py
"""LineIndicator base class for indicators with a single numeric input/output."""

from typing import Callable

from ...entities.bar import Bar
from ...entities.quote import Quote
from ...entities.trade import Trade
from ...entities.scalar import Scalar
from .output import Output


class LineIndicator:
    """Provides update_scalar, update_bar, update_quote and update_trade
    methods for indicators that take a single numeric input and produce
    output via a core update function.

    Concrete indicators should inherit from this class and pass their
    update method to the constructor.
    """

    def __init__(
        self,
        mnemonic: str,
        description: str,
        bar_func: Callable[[Bar], float],
        quote_func: Callable[[Quote], float],
        trade_func: Callable[[Trade], float],
        update_fn: Callable[[float], float],
    ) -> None:
        self.mnemonic = mnemonic
        self.description = description
        self._bar_func = bar_func
        self._quote_func = quote_func
        self._trade_func = trade_func
        self._update_fn = update_fn

    def update_scalar(self, sample: Scalar) -> Output:
        """Updates the indicator given the next scalar sample."""
        value = self._update_fn(sample.value)
        return [Scalar(sample.time, value)]

    def update_bar(self, sample: Bar) -> Output:
        """Updates the indicator given the next bar sample."""
        return self.update_scalar(Scalar(sample.time, self._bar_func(sample)))

    def update_quote(self, sample: Quote) -> Output:
        """Updates the indicator given the next quote sample."""
        return self.update_scalar(Scalar(sample.time, self._quote_func(sample)))

    def update_trade(self, sample: Trade) -> Output:
        """Updates the indicator given the next trade sample."""
        return self.update_scalar(Scalar(sample.time, self._trade_func(sample)))
